File: src/main/java/com/uap/control_tickets/services/interfaces/EstudianteService.java
package com.uap.control_tickets.services.interfaces;

import com.uap.control_tickets.dto.estudiante.EstudianteDetalleDto;
import com.uap.control_tickets.dto.estudiante.EstudianteDto;
import com.uap.control_tickets.dto.estudiante.ImportacionResultadoDto;
import com.uap.control_tickets.dto.estudiante.PrevisualizacionCsvDto;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/** Contrato del CRUD de Estudiante + carga masiva por CSV. */
public interface EstudianteService {
    List<EstudianteDetalleDto> listar();
    EstudianteDetalleDto obtener(Long idEstudiante);
    EstudianteDetalleDto crear(EstudianteDto dto);
    void eliminar(Long idEstudiante);

    /** Carga masiva desde un archivo CSV (nombre,paterno,materno,ci,ru,facultad,carrera). */
    ImportacionResultadoDto importarCsv(MultipartFile archivo);

    /** Lee el CSV y reporta que pasaria al importarlo, sin escribir en la base. */
    PrevisualizacionCsvDto previsualizarCsv(MultipartFile archivo);

    /** Facultades y carreras distintas (para la carga masiva al biométrico). */
    List<String> facultades();

    List<String> carreras();
}

File: frontend/src/api/estudiante.service.ts
// Capa de API de Estudiantes (CRUD + importación CSV).
import http from '@/api/http'
import type { EstudianteDetalleDto, EstudianteDto, ImportacionResultadoDto, PrevisualizacionCsvDto } from '@/types/estudiante.type'

export function listarEstudiantes() {
  return http.get<EstudianteDetalleDto[]>('/estudiantes/listar').then((r) => r.data)
}

export function crearEstudiante(dto: EstudianteDto) {
  return http.post<EstudianteDetalleDto>('/estudiantes/crear', dto).then((r) => r.data)
}

export function eliminarEstudiante(idEstudiante: number) {
  return http.delete('/estudiantes/eliminar', { params: { idEstudiante } })
}

// Importación masiva: se envía el archivo como multipart/form-data.
export function importarEstudiantesCsv(archivo: File) {
  const fd = new FormData()
  fd.append('archivo', archivo)
  return http.post<ImportacionResultadoDto>('/estudiantes/importar', fd).then((r) => r.data)
}

/** Lee el CSV en el servidor y devuelve que pasaria al importarlo (no guarda nada). */
export function previsualizarCsv(archivo: File) {
  const fd = new FormData()
  fd.append('archivo', archivo)
  return http
    .post<PrevisualizacionCsvDto>('/estudiantes/previsualizar', fd)
    .then((r) => r.data)
}

/** Facultades y carreras distintas (para la carga masiva al biométrico). */
export function listarFacultades() {
  return http.get<string[]>('/estudiantes/facultades').then((r) => r.data)
}

export function listarCarreras() {
  return http.get<string[]>('/estudiantes/carreras').then((r) => r.data)
}
